### DISCUSSION TITLE
how to old data see refresh page  ?

### DISCUSSION THREAD

**Discussion by productdevbook** [iWeltAG/pinia-cached-store#3] [Q&A]
```
onMounted(() => {
  pizza.pizzaName;
});
``` 
example refresh page dont see but local storege see data

vue3 + pinia 2

**Comment by yrd** [comment]
I'm not quite sure what you are trying to achieve. `pizza.pizzaName` is a no-op and won't actually do anything. Could you maybe provide a more complete example?

**Comment by productdevbook** [comment]
```
const pi = usePiStore();
onMounted(() => {
 console.log(pi.value)
});
``` 
index.vue 

refresh page all data empty dont see 

it saves the data but when I refresh the page the value is now empty.
```
pi.$load({ value: 99999 });
```
etc save button and refresh page dont see value 9999 empty

**Reply by yrd** [reply]
Okay, a few things here:

- First, the `piStore` doesn't actually take a `value` parameter. Instead, you provide a number of iterations for the algorithm that calculates π. This number basically states how exact you want your result willl be. You will never exactly get the right value, but the more iterations you run the more exact the `.value` result will be.
- Second, the store is asynchronous. You should await it when loading. I have opened an issue (#4) for making this pardigm optional.

The updated example code should look something like this:

```ts
onMounted(async () => {
  // Feel free to play around with this parameter here and get different results:
  await pi.$load({ iterations: 99999 });
  console.log(pi.value);
});
```

In general, the example is supposed to show that you don't actually need to fetch remote data with a cached store but can also use them to calculate the result from some expensive caluation (in this case, approximating π). @productdevbook Would you mind testing with the updated code?

**Comment by yrd** [comment]
Just saw that the documentation wasn't actually up to date and still more or less referred to the synchronous `$load` implementations. This has been updated now, see the readme on the main branch.

**Reply by productdevbook** [reply]
iam testing in 10 minutes

**Reply by productdevbook** [reply]
no some problem

**Reply by productdevbook** [reply]
local storege keeps data but does not fetch it.

**Reply by productdevbook** [reply]
```
 onMounted(async () => {
        if (await missingBookStore.book.isbn) {
          setValues({
            ...(await missingBookStore.book),
          })
        }
 ```

**Reply by yrd** [reply]
Without knowing the details of your implementation, I'm not sure you are using promises correctly. *Only* `.$load()` is async, afterwards (once data is loaded) data should be available normally. In you're example you are awaiting both the `.book` and `.book.isbn` properties, which seems a bit off to me...
It might be easier if you could provide some sort of minimal working example project to reproduce the error?

**Reply by productdevbook** [reply]
```
import {defineCachedStore} from 'pinia-cached-store'
import {CreateBookInput} from '~/api'

export const newMissingBookStore = defineCachedStore({
  id: 'newMissingBook',
  state: () => ({
    book: <CreateBookInput>{},
    isbn: '',
    step: 1,
  }),
  async refresh(book: CreateBookInput) {
    this.book = book
  },
  caching: {
    keyPrefix: 'newBook',
  },
})
```
on send form button

```
    const onSubmit = handleSubmit(async (values) => {
        missingBookStore.$load({
          ...missingBookStore.book,
          ...values,
        })

``` 

and refresh page 

```
 onMounted(async () => {
        if (await missingBookStore.book.isbn) {
          setValues({
            ...(await missingBookStore.book),
          })
        }
 ```
 
 dont see data

**Reply by yrd** [reply]
I'm guessing `setValues` will update your view? What type does `CreateBookInput.book.isbn` have? If it isn't a promise, you can't await it (I'm not even sure what awaiting the property would mean). Try something like this:

``` ts
onMounted(async () => {
  if (missingBookStore.book.isbn) {
    setValues({ ...missingBookStore.book });
  }
});

**Reply by productdevbook** [reply]
`CreateBookInput.book.isbn` string

**Reply by yrd** [reply]
Then you definitely shouldn't be awaiting it. And by the way: are you calling `.$load()` anywhere other that when sending the form? If not, the store will be empty on mount (as intended). You will need to call the load function on mount to populate data from the cache.

**Reply by productdevbook** [reply]
no only send button $load used. How to used onMounted $load ?

**Reply by yrd** [reply]
You need to pass the same arguments that you used in the other `$load` call. So if you called it with something like `{ isbn: '1234' }` from the submit handler it needs to be called again with `{ isbn: '1234' }` when mounting the component. If there is something in the cache for that input object (not the ISBN value, but the whole object is used as a key) that will be loaded. Otherwise, your `refresh` will be called.

**Reply by yrd** [reply]
Note I think you're misunderstanding the concept here: a cached store should be used to cache some predefined but expensive-to-get state. The whole thing is not yet suitable for storing arbitrary data and automatically writing it to local storage. I have opened an issue for this: #5

**Reply by productdevbook** [reply]
okay. maybe etc 
`
 caching: {
    keyPrefix: 'newBook',
  },
`
etc used
```
$load({keyCache: 'newBook' })
```
and load newBook only onMounted. good feature. 
It would be great to recall the one I gave the key to. Because the user is constantly typing on the same form or resetting for new book it. The point is that if the internet breaks down, coming internet  it doesn't wait to be filled in again.

**Reply by yrd** [reply]
So you basically want to cache a user's input? I guess that would be possible with a new syntax once it's implemented...

**Reply by productdevbook** [reply]
Yes, when filling out a form, we always want to send a cache to the back. If the form is successfully added to the database, we will delete the cache. Because the form is long.

maybe user internet connect lose and refresh page and some data see

**Reply by yrd** [reply]
Yeah, that kind of stuff wan't really in scope when we designed the API. It might be easier to use something like this: https://github.com/robinvdvleuten/vuex-persistedstate

**Reply by productdevbook** [reply]
vuex-persistedstate dont pinia used. I love used pinia :D

**Reply by yrd** [reply]
Oh, yeah - that plugin is for Vuex. But there [are](https://www.npmjs.com/search?q=pinia%20persist) a bunch of them for Pinia as well.

Once #5 is implemented, you will be able to do it with `pinia-cached-store` as well, though.

**Reply by productdevbook** [reply]
I'm so glad to hear that. I'm waiting for good news.

**Reply by yrd** [reply]
This should now be released as `pinia-cached-store@0.5.0`.
